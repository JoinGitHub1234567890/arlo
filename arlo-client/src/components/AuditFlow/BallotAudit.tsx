import React from 'react'

interface Props {
  review: () => void
}

const BallotAudit: React.FC<Props> = ({ review }: Props) => {
  return <p>Auditing Ballot</p>
}

export default BallotAudit
