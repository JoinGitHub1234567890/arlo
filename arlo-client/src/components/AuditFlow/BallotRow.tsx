import styled from 'styled-components'

const BallotRow = styled.div`
  display: flex;
  justify-content: flex-start;
  margin: 20px 0;

  .ballot-side {
    width: 200px;
    padding: 20px 0;
  }
  .ballot-main {
    width: 50%;
    padding: 20px;
  }
`

export default BallotRow
