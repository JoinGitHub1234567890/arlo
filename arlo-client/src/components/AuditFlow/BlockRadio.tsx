import React from 'react'
import styled from 'styled-components'
import { IBallot } from '../../types'

const Block = styled.label`
  &.bp3-control.bp3-radio {
    display: inline-block;
    position: relative;
    margin-bottom: 20px;

    .bp3-control-indicator {
      border-radius: unset;
      width: 10em;
      height: 3em;

      &::before {
        display: none;
      }
      .radio-text {
        position: absolute;
        top: 50%;
        transform: translateY(-50%);
        width: 100%;
        padding: 5px;
        text-align: center;
      }
    }
  }
`

interface IProps {
  name?: string
  value: Exclude<IBallot['vote'], null>
  handleChange?: (e: React.ChangeEvent<HTMLInputElement>) => void
  locked?: boolean
  className?: string
  checked?: boolean
}

const TEXT = {
  YES: 'Yes/For',
  NO: 'No/Against',
  NO_CONSENSUS: 'No audit board consensus',
  NO_VOTE: 'Blank vote/no mark',
}

const BlockRadio = ({
  name = '',
  value,
  handleChange,
  locked,
  className,
  checked,
}: IProps) => (
  <Block className={`${className} bp3-control bp3-radio`}>
    {locked ? (
      <input type="radio" name={name} value={value} disabled checked />
    ) : (
      <input
        type="radio"
        name={name}
        data-testid={value}
        value={value}
        onChange={handleChange}
        checked={checked}
      />
    )}
    <span className="bp3-control-indicator">
      <span className="radio-text">{TEXT[value]}</span>
    </span>
  </Block>
)

export default BlockRadio
