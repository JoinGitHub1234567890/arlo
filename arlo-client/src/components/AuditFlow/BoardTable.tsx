import React from 'react'
import styled from 'styled-components'
import { Table, Column, Cell } from '@blueprintjs/table'
import { H1, Button } from '@blueprintjs/core'
import { Link } from 'react-router-dom'
import { IAuditBoard, IBallot } from '../../types'

const RightWrapper = styled.div`
  display: flex;
  flex-direction: row-reverse;
  margin: 20px 0;
`

const PaddedCell = styled(Cell)`
  padding: 3px 5px;
`

const RightButton = styled(Button)`
  float: right;
`

const ActionWrapper = styled.div`
  margin-bottom: 20px;
  .bp3-checkbox {
    display: inline-block;
    margin-left: 20px;
  }
`

interface IProps {
  setIsLoading: (arg0: boolean) => void
  isLoading: boolean
  board: IAuditBoard
  url: string
}

const KEYS: ('tabulator' | 'batch' | 'position' | 'status')[] = [
  'tabulator',
  'batch',
  'position',
  'status',
]

const STATUSES: { [key: string]: string } = {
  AUDITED: 'Audited',
  NOT_AUDITED: 'Not Audited',
}

const BoardTable: React.FC<IProps> = ({ board, url }: IProps) => {
  const renderCell = (rI: number, cI: number) => {
    if (board.ballots) {
      const row: IBallot = board.ballots[rI]
      if (!KEYS[cI]) {
        return <PaddedCell>{board.name}</PaddedCell>
      } else if (STATUSES[row[KEYS[cI]]]) {
        const action =
          row[KEYS[cI]] === 'AUDITED' ? (
            <RightButton small>Re-audit</RightButton>
          ) : null
        return (
          <PaddedCell>
            <>
              {STATUSES[row[KEYS[cI]]]} {action}
            </>
          </PaddedCell>
        )
        // wrapping content in React.Fragment to avoid unexpected props being passed to dom elements and throwing warnings
      } else {
        return <PaddedCell>{row[KEYS[cI]]}</PaddedCell>
      }
    } else {
      return <PaddedCell loading />
    }
  }

  const columnWidths = (): (number | null)[] => {
    const container = document.getElementsByClassName(
      'board-table-container'
    )[0]
    if (!container) return Array(5).fill(null)
    const containerSize = container.clientWidth
    const colWidth = (containerSize - 31) / 5
    return Array(5).fill(colWidth)
  }

  const roundComplete =
    board.ballots && board.ballots.every(b => b.status === 'AUDITED')

  return (
    <div className="board-table-container">
      <H1>{board.name}: Ballot Cards to Audit</H1>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
        commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
        velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint
        occaecat cupidatat non proident, sunt in culpa qui officia deserunt
        mollit anim id est laborum.
      </p>
      <RightWrapper>
        {roundComplete ? (
          <Button intent="primary">Review Complete - Finish Round</Button>
        ) : (
          <Link
            to={url + '/round/1/ballot/1'}
            className="bp3-button bp3-intent-primary"
          >
            Start Auditing
          </Link>
        )}
      </RightWrapper>
      <ActionWrapper>
        {!roundComplete && (
          <>
            <Button intent="primary">Download Ballot List as CSV</Button>
          </>
        )}
      </ActionWrapper>
      <Table numRows={10} defaultRowHeight={30} columnWidths={columnWidths()}>
        <Column key="tabulator" name="Tabulator" cellRenderer={renderCell} />
        <Column key="batch" name="Batch" cellRenderer={renderCell} />
        <Column
          key="position"
          name="Record/Position"
          cellRenderer={renderCell}
        />
        <Column key="status" name="Status" cellRenderer={renderCell} />
        <Column name="Audit Board" cellRenderer={renderCell} />
      </Table>
    </div>
  )
}

export default BoardTable
