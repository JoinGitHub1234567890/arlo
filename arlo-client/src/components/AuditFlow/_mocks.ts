import { IAuditBoard } from '../../types'

export const dummyBoard: IAuditBoard[] = [
  {
    id: '123',
    name: 'Audit Board #1',
    members: [],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: 'Precinct 13',
        position: '313',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: 'Precinct 13',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
    ],
  },
]

export default {
  dummyBoard,
}
