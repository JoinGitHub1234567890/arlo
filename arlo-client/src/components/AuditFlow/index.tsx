import React from 'react'
import styled from 'styled-components'
import { AuditFlowParams } from '../../types'

const P = styled.p`
  margin-top: 100px;
`

interface Props {
  match: {
    params: AuditFlowParams
  }
}

const AuditFlow: React.FC<Props> = ({
  match: {
    params: { token },
  },
}: Props) => {
  return <P>Audit: {token}</P>
}

export default AuditFlow
