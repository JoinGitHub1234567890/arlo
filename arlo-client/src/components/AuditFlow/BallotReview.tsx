import React from 'react'
import { H3, Divider, Button } from '@blueprintjs/core'
import styled from 'styled-components'
import { IReview } from '../../types'
import { BallotRow, FormBlock, ProgressActions } from './Atoms'
import FormButton from '../Form/FormButton'
import BlockRadio from './BlockRadio'

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 20px 0;
`

const SingleBlockRadio = styled(BlockRadio)`
  &.bp3-control.bp3-radio {
    margin: 0;
  }
`

interface IProps {
  contest: string
  goAudit: () => void
  review: IReview
}

const BallotReview: React.FC<IProps> = ({
  contest,
  goAudit,
  review: { vote, comment },
}: IProps) => {
  /* eslint-disable no-console */
  const handleSubmit = () => console.log(vote, comment)
  return (
    <BallotRow>
      <div className="ballot-side"></div>
      <div className="ballot-main">
        <FormBlock>
          <H3>{contest}</H3>
          <Divider />
          <Wrapper>
            <SingleBlockRadio value={vote} locked />
            <Button onClick={goAudit} icon="edit" minimal>
              Edit
            </Button>
          </Wrapper>
          <p>COMMENT: {comment}</p>
        </FormBlock>
        <ProgressActions>
          <FormButton type="submit" onClick={handleSubmit} intent="success">
            Submit &amp; Next Ballot
          </FormButton>
          <Button onClick={goAudit} minimal>
            Back
          </Button>
        </ProgressActions>
      </div>
    </BallotRow>
  )
}

export default BallotReview
