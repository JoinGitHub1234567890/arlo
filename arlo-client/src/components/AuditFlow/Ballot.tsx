import React, { useState } from 'react'
import { H1, H3, Callout, H4, Divider } from '@blueprintjs/core'
import styled from 'styled-components'
import BallotAudit from './BallotAudit'
import BallotReview from './BallotReview'
import { AuditBoard } from '../../types'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`

const MainCallout = styled(Callout)`
  width: 400px;
  font-weight: 700;
`

const Info = styled.div`
  display: flex;
  justify-content: flex-start;
  margin: 20px 0;

  .ballot-info {
    width: 200px;
    padding: 20px 0;
  }
  .description {
    width: 50%;
    padding: 20px;
  }
`

interface Props {
  roundId: string
  ballotId: string
  board: AuditBoard
}

const Ballot: React.FC<Props> = ({ roundId, ballotId, board }: Props) => {
  const [auditing, setAuditing] = useState(true)

  if (!board.ballots) return null // TODO handle informatively
  const ballot = board.ballots[Number(ballotId) - 1]
  if (!ballot) return null // TODO handle informatively

  return (
    <Wrapper>
      <H1>{board.name}: Ballot Card Data Entry</H1>
      <H3>Enter Ballot Information</H3>
      <MainCallout icon={null} intent="primary">
        Round {roundId}: auditing ballot {ballotId} of {board.ballots.length}
      </MainCallout>
      <Info>
        <div className="ballot-info">
          <H4>Current ballot:</H4>
          <div>Tabulator: {ballot.tabulator}</div>
          <div>Batch: {ballot.batch}</div>
          <div>Record/Position: {ballot.record}</div>
        </div>
        <Divider />
        <div className="description">
          <H4>Are you looking at the correct ballot?</H4>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in
            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
            pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
            culpa qui officia deserunt mollit anim id est laborum.
          </p>
        </div>
      </Info>
      {auditing ? (
        <BallotAudit review={() => setAuditing(false)} />
      ) : (
        <BallotReview audit={() => setAuditing(true)} />
      )}
    </Wrapper>
  )
}

export default Ballot
