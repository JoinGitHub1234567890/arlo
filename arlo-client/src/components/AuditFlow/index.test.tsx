import React from 'react'
import { render } from '@testing-library/react'
import { routerTestProps } from '../testUtilities'
import AuditFlow from './index'

const routeProps = routerTestProps('/board/:token', {
  electionId: '1',
  token: '123',
})

it('renders correctly', () => {
  const { container } = render(<AuditFlow {...routeProps} />)
  expect(container).toMatchSnapshot()
})
