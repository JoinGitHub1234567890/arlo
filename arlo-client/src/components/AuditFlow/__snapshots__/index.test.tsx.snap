// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <p
    class="sc-bdVaJa dHwYGt"
  >
    Audit: 
    123
  </p>
</div>
`;
